content/automate/ManageIQ/Infrastructure/VM/Retirement/StateMachines/Methods.class/__methods__/check_powered_off.rb

```diff
@@ -2,24 +2,56 @@
 # Description: This method checks to see if the VM has been powered off or suspended
 #
 
-# Get vm from root object
-vm = $evm.root['vm']
-
-if vm
-  power_state = vm.attributes['power_state']
-  ems = vm.ext_management_system
-  $evm.log('info', "VM:<#{vm.name}> on provider:<#{ems.try(:name)} has Power State:<#{power_state}>")
-
-  # If VM is powered off or suspended exit
-
-  if %w(off suspended).include?(power_state)
-    # Bump State
-    $evm.root['ae_result'] = 'ok'
-  elsif power_state == "never"
-    # If never then this VM is a template so exit the retirement state machine
-    $evm.root['ae_result'] = 'error'
-  else
-    $evm.root['ae_result']     = 'retry'
-    $evm.root['ae_retry_interval'] = '60.seconds'
+module ManageIQ
+  module Automate
+    module Infrastructure
+      module VM
+        module Retirement
+          module StateMachines
+            class CheckPoweredOff
+              def initialize(handle = $evm)
+                @handle = handle
+              end
+
+              def main
+                # Get vm from root object
+                vm = @handle.root['vm']
```

Comment:
@billfitzgerald0120 
Can we write a function that returns the vm and raises an error if the vm is missing @handle.root[vm] is nil